evolve m with heff

jackieyao0114 opened the pull request:
Key contribution: added the functionality of solving the magnetization using Landau-Lifshitz_Gilbert equation.

Notes on current version:
1. No Maxwellian fields involved yet. In mathematical formula, H_effective= H_bias, with no contribution from H_Maxwell. However, the calculation of H_Maxwell is implemented already, and can be readily added into H_effective.
2. Because of the lack of traveling waves, the CFL limit on the time step size is lifted for this case, and this is reason that warpx.cfl = 4000 in the input file (Line 19).
2. No iterative correction of M included yet, i.e. the iteration shown in Section 3.2.2 of the pdf document is NOT included.
3. No normalization added to the M calculation yet.

Documentation on the code and demo case: **Section 3.2** . If any question comes up about the unit conversion, please refer to **Section 3.2.3 - Initialization.**

[EvolveM_1st demo.pdf](https://github.com/RevathiJambunathan/WarpX/files/4893239/EvolveM_1st.demo.pdf)

Input file for the demo:

```
# ###############################
####### GENERAL PARAMETERS ######
#################################
max_step = 1000
amr.n_cell = 8 8 8 # number of cells spanning the domain in each coordinate direction at level 0
amr.max_grid_size = 512 # maximum size of each AMReX box, used to decompose the domain
amr.blocking_factor = 8
geometry.coord_sys   = 0
geometry.is_periodic = 1  1  1
geometry.prob_lo     =  -1.5e-6 -1.5e-6 -1.5e-6
geometry.prob_hi     =   1.5e-6  1.5e-6  1.5e-6
amr.max_level = 0
#################################
############ NUMERICS ###########
#################################
warpx.verbose = 0
warpx.use_filter = 0
warpx.cfl = 4000
warpx.do_pml = 0
#algo.em_solver_medium = vacuum # vacuum/macroscopic
algo.em_solver_medium = macroscopic # vacuum/macroscopic
algo.macroscopic_sigma_method = laxwendroff # laxwendroff or backwardeuler
macroscopic.sigma_init_style = "parse_sigma_function" # parse or "constant"
macroscopic.sigma_function(x,y,z) = "0.0"
#macroscopic.sigma = 0.0
macroscopic.epsilon_init_style = "parse_epsilon_function" # parse or "constant"
macroscopic.epsilon_function(x,y,z) = "8.8541878128e-12"
#macroscopic.epsilon = 8.8541878128e-12
macroscopic.mu_init_style = "parse_mu_function" # parse or "constant"
macroscopic.mu_function(x,y,z) = "1.25663706212e-06"
#macroscopic.mu = 1.25663706212e-06
macroscopic.mag_Ms_init_style = "parse_mag_Ms_function" # parse or "constant"
macroscopic.mag_Ms_function(x,y,z) = "1.4e5"
#macroscopic.mag_Ms = 1.4e5 # in unit A/m, equal to 1750 Gauss
#unit conversion: 1 Gauss = (1000/4pi) A/m
macroscopic.mag_alpha_init_style = "parse_mag_alpha_function" 
macroscopic.mag_alpha_function(x,y,z) = "5e-01"
#macroscopic.mag_alpha = 5e-02 # alpha is unitless, typical values range from 1e-3 ~ 1e-5
macroscopic.mag_gamma_init_style = "parse_mag_gamma_function"
macroscopic.mag_gamma_function(x,y,z) = "1.759e11"
#macroscopic.mag_gamma = 1.759e11 # gyromagnetic ratio is constant for electrons in all materials
particles.nspecies = 0
#################################
############ FIELDS #############
#################################
my_constants.pi = 3.14159265359
my_constants.L = 141.4213562373095e-6
my_constants.c = 299792458.
my_constants.wavelength = 1.2e-1
warpx.E_ext_grid_init_style = parse_E_ext_grid_function
warpx.Ez_external_grid_function(x,y,z) = 0.
warpx.Ex_external_grid_function(x,y,z) = 0.
warpx.Ey_external_grid_function(x,y,z) = 0.
#warpx.Ey_external_grid_function(x,y,z) = "1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)"
warpx.B_ext_grid_init_style = parse_B_ext_grid_function
#warpx.Bx_external_grid_function(x,y,z)= "-1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)/c"
warpx.Bx_external_grid_function(x,y,z)= 0.1759.
warpx.By_external_grid_function(x,y,z)= 0.
warpx.Bz_external_grid_function(x,y,z) = 0.
warpx.H_bias_ext_grid_init_style = parse_H_bias_ext_grid_function
warpx.Hx_bias_external_grid_function(x,y,z)= 0.
warpx.Hy_bias_external_grid_function(x,y,z)= 0.
warpx.Hz_bias_external_grid_function(x,y,z)= 3e4 # in A/m, equal to 382 Oersted
#unit conversion: 1 Gauss = 1 Oersted = (1000/4pi) A/m
#calculation of H_bias: H_bias (oe) = frequency / 2.8e6
#warpx.H_bias_ext_grid_init_style = constant
#warpx.Hx_bias_external_grid = 0
#warpx.Hy_bias_external_grid = 0
#warpx.Hz_bias_external_grid = 0
#warpx.M_ext_grid_init_style = parse_M_ext_grid_function
#warpx.Mx_external_grid_function(x,y,z)= "-1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)/c"
#warpx.My_external_grid_function(x,y,z)= 0.
#warpx.Mz_external_grid_function(x,y,z) = 0.
warpx.M_ext_grid_init_style = constant
warpx.M_external_grid = 140000. 0. 0.
# Diagnostics
diagnostics.diags_names = plt
plt.period = 10
plt.diag_type = Full
plt.fields_to_plot = Mx_xface My_xface Mz_xface Mx_yface My_yface Mz_yface Mx_zface My_zface Mz_zface
plt.plot_raw_fields = 0

```
Simulation results compared against a simple LLG Matlab code:
![M_comparison](https://user-images.githubusercontent.com/58234082/86970009-8786dc80-c123-11ea-9ab7-972550414633.png)

Matlab code for validation:

```
clearvars
close all

% This is a 1D FDTD code solving for the FMR of a single ferrite layer
% LLG equation is original nonlinear form
% The DC bias magnetic field is applied along y direction, and the AC
% dynamic field is added on the x and z direction.
%
% The coordinate system is defined as:
%
%  -----------------------------------     ^
%                                          | z 
%
%  -----------------------------------    --> x
%

%% Universal parameters
EP0 = 8.84e-12; % Epsilon, free space permittivity 
MU0 = 4.*pi*1.0E-7; % Mu, free space permeability
CC = 3.0*1.e8; % Speed of light in free space
freq = 1.068e9; % Frequency of the wave, if sinusoidal excitation is used
eta0 = 120*pi;
gamma = 1.759*10^11; % Unit is C/kg

% Define material properties
H0_oe = 382;
Ms_ga = 1750;

%% Set up geometry
% Only the vertical direction is considered
TS = 3.0e-6; % The heigh of an individule layer
NZ = 10; % number of cells you want to divide in Z direction.
DZ = TS/NZ;
TP = 3*1/freq; % Define the Gaussian pulse width in the excitation; 

%% Set up the time steps
% DT  =  1/freq/200; % define FDTD time-step size (not constrained by CFL limit)
Tstep_end = 2000; % total number of time steps
DT_CFL = DZ/CC;
DT = 1.444e-12; % match WarpX setup

Tstep_int = 10; % number of time steps between adjacent record points
Tstep_record = floor(Tstep_end/Tstep_int); % # of recorded data points
Time_axis = Tstep_int:Tstep_int:Tstep_end;
NFFT = 2^nextpow2(Tstep_record); % # of points for FFT
fs = 1/(Tstep_int*DT); % highest frequency included in FFT
f = fs/2*linspace(0,1,(NFFT/2)+1);
f_GHz = f/1e9;

Tstep_start = 1;
Tstep_zeros = 10000; % # of zero points added after the pulse dies out
Tstep_ext = Tstep_record+Tstep_zeros; % total length of time-domain signal after extending time window
NFFT_ext = 2^nextpow2(Tstep_ext-Tstep_start+1); % new FFT point # after extension
f_ext = fs/2*linspace(0,1,(NFFT_ext/2)+1);
f_ext_GHz = f_ext/1e9;

%% Analytical calculation
H0 = H0_oe/(4*pi*10^-3); % unit is A/m
Ms = Ms_ga/(4*pi*10^-3); % unit is tesla, 4*pi*Ms = 1750 gauss
wm = 2*pi*2.8e6*Ms_ga;  % saturation magnetization
w0_0 = 2*pi*2.8e6*H0_oe;  % external biasing field 
w = 2*pi*f_ext;
wr = sqrt(w0_0*(wm+w0_0)); % FMR estimation based on Kittel's equation

% delta_H = 10; % line width in the unit of Oesterd
% alpha = delta_H*2.8e6*2*pi/wr/2;  % dampling constant
% alpha = delta_H*2.8e6/9.6e9/2; % dampling constant
% w0 = w0_0+1i*alpha*w; % Complex FMR of bulk due to damping
% su_complex = wm*(wm+w0)./(w0.*(wm+w0)-w.^2); % complex susceptability of the film
% mu_p = 1+real(su_complex); % real part of the permeability
% mu_d = -imag(su_complex); % imaginary part of the permeability
% tand = mu_d./mu_p; % magnetic loss tangent

%% Time domain update

% Initialize field components;
Hx = 0; 
Hy = 0; 
Hz = 0; 
% Mx = 1715*(1e3/4/pi); 
% My = 300*(1e3/4/pi); 
% Mz = 0; 
Mx = Ms; 
My = 0; 
Mz = 0; 

% Initialize field matrices
Hx_obs = zeros(1,Tstep_record);
Hy_obs = zeros(1,Tstep_record);
Hz_obs = zeros(1,Tstep_record);
Mx_obs = zeros(1,Tstep_record);
My_obs = zeros(1,Tstep_record);
Mz_obs = zeros(1,Tstep_record);

alpha = 0.5;
Prefactor1 = - gamma*MU0 * DT ; % Note DT is 100 larger than what it is supposed to be
Prefactor2 = Prefactor1 * alpha / Ms;

for Tstep = 1 : Tstep_end
    
    TT = Tstep*DT;
    
    if TT < 0e-9
        Hz = 0;
    else
        Hz = H0_oe*(1e3/4/pi);
    end

    deltaMx = Prefactor1 .* (My .* Hz - Mz .* Hy) + Prefactor2 .* (Mx .* (My .* Hy + Mz .* Hz) - My.^2 .* Hx - Mz.^2 .* Hx);
    deltaMy = Prefactor1 .* (Mz .* Hx - Mx .* Hz) + Prefactor2 .* (My .* (Mz .* Hz + Mx .* Hx) - Mz.^2 .* Hy - Mx.^2 .* Hy);
    deltaMz = Prefactor1 .* (Mx .* Hy - My .* Hx) + Prefactor2 .* (Mz .* (Mx .* Hx + My .* Hy) - Mx.^2 .* Hz - My.^2 .* Hz);
    
    Mx = Mx + deltaMx;
    My = My + deltaMy;
    Mz = Mz + deltaMz;
    
    mag = sqrt(Mx .* Mx + My .* My + Mz .* Mz);
    Mx = Mx ./ mag * Ms;
    My = My ./ mag * Ms;
    Mz = Mz ./ mag * Ms;
    
    if rem(Tstep, Tstep_int) == 0
        NN = floor(Tstep/Tstep_int);
        Mx_obs(NN) = Mx;
        My_obs(NN) = My;
        Mz_obs(NN) = Mz;
    end
    
end

figure
plot(Time_axis*DT*1e9, Mx_obs/(1e3/4/pi), 'k', 'Linewidth', 2)
hold on
plot(Time_axis*DT*1e9, My_obs/(1e3/4/pi), ':r', 'Linewidth', 4)
plot(Time_axis*DT*1e9, Mz_obs/(1e3/4/pi), '-.b', 'Linewidth', 2)

legend ('Mx','My','Mz')
xlabel('Time [ns]');
ylabel('4\piM [Gauss]'); 
title( ['Magnetization magnitude, \alpha = ', num2str(alpha)]);
```

jackieyao0114 committed:
fixing conflicts in Source/Evolve/WarpXEvolve.cpp

jackieyao0114 committed:
fixed conflicts in Source/Evolve/WarpXEvolve.cpp with FillBoundaryM

jackieyao0114 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into evolveM_dev

jackieyao0114 committed:
added macroscopic material properties to EvolveM

jackieyao0114 committed:
added MacroscopicEvolveM.cpp to the Make.package file; fixed the way of calling Ms_arr and alpha_arr in MacroscopicEvolveM.cpp

jackieyao0114 committed:
trying to merge Reva's newest evolveM_dev; didn't compile; will revert to previous commit of evolveM_parse anyways

jackieyao0114 committed:
Revert "trying to merge Reva's newest evolveM_dev; didn't compile; will revert to previous commit of evolveM_parse anyways"

This reverts commit 1e2f9ce27d152beadf7c800f3e2fa7cf3e1cd1f6, reversing
changes made to 42b8a8ff8c67608e661520e058842df4e8612828.

jackieyao0114 committed:
mergered Reva's latest evolveM_dev

jackieyao0114 committed:
added H_Maxwell, without H_bias; defined H multifab; mag_gamma defined as PhysConst

jackieyao0114 committed:
identical to previous commit but to add more notes: nothing was changed in MacroscopicEvolveE yet, i.e. we should have used H in updating E but current version uses B. Keep in mind to fix this

jackieyao0114 committed:
changed mag_gamma into user defined multifab

jackieyao0114 committed:
pull from Reva's evolveM_dev with recent ifdef mag

jackieyao0114 committed:
Revert "pull from Reva's evolveM_dev with recent ifdef mag"

This reverts commit cdef9dbd84ca197981496a6f181c8d61c51b72ad, reversing
changes made to 0b0e1e3dbfe80a59d910b6f771e8bc32d001733d.

jackieyao0114 committed:
added H_bias parser and read from input file inputs_3d.evolveM; but not used in MacroscopicEvolveM.cpp yet

jackieyao0114 committed:
using H_bias in MacroscopicEvolveM.cpp; defined a new local variable H_eff to make M evolving equations cleaner

jackieyao0114 committed:
fixed typo in WarpXInitData.cpp Line 498-500; added read in M in the same file but commented out

jackieyao0114 committed:
changed H_Maxwell into local variable and added functions to calculate H_Maxwell; fixed one typo in MacroscopicEvolveE.cpp Line 53

jackieyao0114 committed:
changed EvolveB to be full time step implementation

jackieyao0114 committed:
added Examples/Tests/Macroscopic_Maxwell/inputs_3d.convergence

jackieyao0114 committed:
changed USE_GPU = TRUE in GNUmakefile; changed input M with constant flag in Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM

jackieyao0114 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into evolveM_withHeff

jackieyao0114 committed:
fixed conflicts when merging from Reva/evolveM_dev

jackieyao0114 committed:
fixed conflicts when pulling from Reva/evolveM_dev; previous commit 0c02a5e20e1607faec0430efbd239218ec8bb3eb caused problem

jackieyao0114 committed:
adding files after conflicts

jackieyao0114 committed:
fixed conflicts after removing 0c02a5e20e1607faec0430efbd239218ec8bb3eb

jackieyao0114 committed:
changed USE_GPU = FALSE in GNUmakefile

jackieyao0114 committed:
added #endif back to the end of #ifdef WARPX_MAG_LLG

jackieyao0114 committed:
added Source/Diagnostics/FlushFormats/FlushFormatSensei* from Reva/evolveM_dev

jackieyao0114 committed:
update Source/Diagnostics/FlushFormats/FlushFormat* from Reva/evolveM_dev

jackieyao0114 committed:
updated /Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolvem to validate the way of inputting M with flag constant

jackieyao0114 committed:
git pull from Reva/evolveM_dev

jackieyao0114 committed:
fixed conflicts in  Source/WarpX.H

jackieyao0114 committed:
checked out .github/workflows/macos.yml from Reva/evolveM

jackieyao0114 committed:
check out .github/workflows/ubuntu.yml from Reva/evolveM_dev

jackieyao0114 committed:
checked out Docs/Doxyfile from Reva/evolveM_dev

jackieyao0114 committed:
checked out Docs/source/conf.py from Reva/evolveM_dev

jackieyao0114 committed:
Docs/source/developers/developers.rst

jackieyao0114 committed:
Docs/source/developers/dimensionality.rst

jackieyao0114 committed:
Docs/source/developers/qed.rst

jackieyao0114 committed:
Docs/source/libensemble/libensemble.rst

jackieyao0114 committed:
Docs/source/running_cpp/parameters.rst

jackieyao0114 committed:
removed Examples/Tests/Macroscopic_Maxwell/inputs_3d.convergence

jackieyao0114 committed:
removed Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM

jackieyao0114 committed:
Examples/Tests/subcycling/inputs_2d

jackieyao0114 committed:
changed GNUmakefile to match Reva/evolveM_dev exactly

jackieyao0114 committed:
LICENSE.txt

jackieyao0114 committed:
Python/setup.py

jackieyao0114 committed:
Source/Diagnostics/FieldIO.H

jackieyao0114 committed:
Source/Diagnostics/FieldIO.cpp

jackieyao0114 committed:
Source/Diagnostics/FlushFormats/Make.package

jackieyao0114 committed:
Source/Diagnostics/FullDiagnostics.H

jackieyao0114 committed:
Source/Diagnostics/FullDiagnostics.cpp

jackieyao0114 committed:
Source/Diagnostics/WarpXIO.cpp

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/AnyFFT.H

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/Make.package

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/SpectralFieldData.H

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/SpectralKSpace.cpp

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/WrapCuFFT.cpp

jackieyao0114 committed:
Source/FieldSolver/SpectralSolver/WrapFFTW.cpp

jackieyao0114 committed:
Source/Particles/Pusher/UpdateMomentumBoris.H

jackieyao0114 committed:
Source/Particles/Pusher/UpdateMomentumBorisWithRadiationReaction.H

jackieyao0114 committed:
Source/Particles/Pusher/UpdateMomentumHigueraCary.H

jackieyao0114 committed:
Source/Particles/Pusher/UpdateMomentumVay.H

jackieyao0114 committed:
Source/Particles/Pusher/UpdatePosition.H

jackieyao0114 committed:
Source/Utils/WarpX_Complex.H

jackieyao0114 committed:
checked out Source/Evolve/WarpXEvolve.cpp from Reva/evolveM_dev

jackieyao0114 committed:
modified  Source/Evolve/WarpXEvolve.cpp: renamed MacroscopicEvolveM functions; moved EvolveB to front of MacroscopicEvolveM

jackieyao0114 committed:
renamed MacroscopicEvolveM; added ifdef Mag_LLG

jackieyao0114 committed:
checked out Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveE.cpp from Reva/evolveM_dev

jackieyao0114 committed:
Source/Initialization/WarpXInitData.cpp from Reva/evolveM_dev

jackieyao0114 committed:
removed unnecessary insitu leftover functions, synced with evolveM_dev

jackieyao0114 committed:
removed residual insitu in WarpX.H

jackieyao0114 committed:
clear up insitu residual and updated WarpX::InitializeExternalFieldsOnGridUsingParser to sync Reva/evolveM_dev

jackieyao0114 committed:
added ifdef mag_llg in MacroscopicProperties.H

jackieyao0114 committed:
added ifdef mag_llg tag in Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H

jackieyao0114 committed:
added ifdef mag_llg tag to Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

jackieyao0114 committed:
added ifdef mag_llg to Source/FieldSolver/FiniteDifferenceSolver/Make.package

jackieyao0114 committed:
added ifdef mag_llg tag to WarpX.H

jackieyao0114 committed:
fixed white space in WarpX.H

jackieyao0114 committed:
added more ifdef mag_llg tags to WarpX.cpp

jackieyao0114 committed:
checked out upstream Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp to fix } issue

jackieyao0114 committed:
fixed } issue

jackieyao0114 committed:
checked out inputs_3d.evolveM from origin/evolveM_dev

jackieyao0114 committed:
added EvolveM functions back for future applications such as nuclear spins

ajnonaka commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -115,7 +117,9 @@ WarpX::Evolve (int numsteps)
-            FillBoundaryM(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
+            if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+                FillBoundaryM(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
+            }
```
should we include:
`else {
amrex::Abort("unsupported em_solver_medium for M field");
}
`

jackieyao0114 replied:
Sure! I added it just now.

jackieyao0114 replied:
I deleted the part about `em_solver_medium` per the previous conversation

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/EvolveM.cpp:
```diff
@@ -56,11 +56,12 @@ void FiniteDifferenceSolver::EvolveM (
+
+        Real constexpr cons1 = -mag_gamma; // should be mu0*mag_gamma, mu0 is absorbed by B used in this case
+        Real constexpr cons2 = -cons1*mag_alpha/Ms; // factor of the second term in scalar LLG
 
```
Is EvolveM() ever supposed to be used, or only MacroscopicEvolveM()?
This function as is doesn't allow you to change the material properties without changing the source code.
MacrosopicProperties.H has m_mag_Ms, m_mag_alpha, and m_mag_gamma; should we just use those?

jackieyao0114 replied:
EvolveM() is not used for now. It is kept here just  so that in the future we are interested in applications like nuclear spin resonances which mostly doesn't involve varying macroscopic material properties. I'll fix EvolveM after this PR is approved. Or I can delete it for now if preferred.

ajnonaka replied:
I suggest fixing it or adding an Abort() in there so nobody inadvertently runs it incorrectly.

jackieyao0114 replied:
I added the amrex::Abort in the beginning of EvolveM.cpp

ajnonaka replied:
I would add one more abort by these definitions of mag_alpha, etc., saying that the code should be rewritten so these can be specified in the inputs file (so we won't forget).

jackieyao0114 replied:
I think I got your idea (and I totally agree that we should have some remark about the user-defined parameters), but I'm not quite sure if I got how you wanted it to be fixed. Could you clarify or we can have a quick meeting tomorrow?

RevathiJambunathan replied:
I think if this function is not being used, it should be removed. it does not really do anything different from what MacroscopicEvolveM does? or is it funtionally different? Plus that values are hard-coded here. If really needed.

ajnonaka replied:
Maybe add something like: "amrex::Abort("Macroscopic parameters in EvolveM.cpp should be set via inputs file");"
Just don't let the code run and put an Abort stating it still needs to be fixed.  Up to you if you think this should be added.

ajnonaka commented on Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM:
```diff
@@ -0,0 +1,88 @@
+# Diagnostics
+diagnostics.diags_names = diag1
+diag1.period = 20
+diag1.diag_type = Full
```
add:
diag1.fields_to_plot = Ex Ey Ez Bx By Bz Mx_xface Mx_yface Mx_zface My_xface My_yface My_zface Mz_xface Mz_yface Mz_zface

jackieyao0114 replied:
I added M plots for now. Thank you for the suggestion.

jackieyao0114 committed:
fix conflict in Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H

jackieyao0114 committed:
update inputs_3d.evolveM

jackieyao0114 committed:
added abort command if medium is not macroscopic in WarpXEvolve.cpp

ruohai0925 committed:
USE_LLG=TRUE

jackieyao0114 committed:
fix if medium invalid statement

ruohai0925 committed:
Merge branch 'debug_evolveM_withHeff' of https://github.com/jackieyao0114/WarpX into debug_evolveM_withHeff_jordan

ruohai0925 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into debug_evolveM_withHeff_jordan

ruohai0925 committed:
push for debugging

jackieyao0114 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into debug_evolveM_withHeff

ruohai0925 committed:
Fix the third bug

ruohai0925 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into debug_evolveM_withHeff_jordan

jackieyao0114 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into debug_evolveM_withHeff

ruohai0925 committed:
Before deleting the print out

ruohai0925 committed:
Delete some print out staff

ruohai0925 committed:
Tries to merge with reva's evolveM_dev

jackieyao0114 committed:
fixed initializing at boundaries; fix face_ave_face

jackieyao0114 committed:
USE_LLG = TRUE

jackieyao0114 committed:
pulled from upstream evolveM_dev; unmerged

jackieyao0114 committed:
fixed conflicts from upstream evolveM_dev in WarpX.H

ruohai0925 committed:
Fix the conflict in the WarpXInitData.cpp

ruohai0925 committed:
Fix the Source/WarpX.H

jackieyao0114 committed:
fixed conflicts in Source/Initialization/WarpXInitData.cpp

ruohai0925 committed:
Delete some unnecessary staffs

ruohai0925 committed:
Modify the indentation

ruohai0925 committed:
Push again

ruohai0925 committed:
Fix the indentation

ruohai0925 committed:
Push again

jackieyao0114 committed:
Merge branch 'debug_evolveM_withHeff' into debug_evolveM_withHeff_jordan

jackieyao0114 committed:
fixed gilbert damping factor

jackieyao0114 committed:
Merge pull request #3 from ruohai0925/debug_evolveM_withHeff_jordan

Fix some bugs of branch jackie/debug_evolveM_withHeff

jackieyao0114 committed:
fixed the negative sign in front of gilbert damping term; fixed mag_gamma value in inputs

jackieyao0114 committed:
deleted amrex::Print in Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp; updated Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM

jackieyao0114 committed:
switched back Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp to included only H_bias, no H_Maxwell, as evolveM_HtoE is not ready yet

jackieyao0114 committed:
Merge branch 'evolveM_dev' of https://github.com/RevathiJambunathan/WarpX into debug_evolveM_withHeff

jackieyao0114 committed:
added Abort in Source/FieldSolver/FiniteDifferenceSolver/EvolveM.cpp to avoid accidental call of this function

ajnonaka commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -355,15 +359,18 @@ WarpX::OneStep_nosub (Real cur_time)
+
         EvolveB(0.5*dt[0]); // We now have B^{n+1/2}
+        FillBoundaryB(guard_cells.ng_FieldSolver, IntVect::TheZeroVector());
 
```
For the coupled M + E + B evolution, we are eventually :) going to want EvolveB to occur before EvolveM (since the 2nd-order evolveM relies on B at the 'old' and 'new' times).  You should be able to switch it now since we are doing M-only tests.

Also we will need the notion of storing old-time information about B (either the B field itself, or the old-time 'source' needed for the 2nd-order M update; but that can be added later).

jackieyao0114 replied:
That's right! 

EvolveM now happens after EvolveB. You can see two EvolveB(0.5*dt[0]) occurs right before two MacroscopicEvolveM(0.5*dt[0]) now. 

But yes, the notion of storing the previous B will be needed and let's add it after this PR.

ajnonaka commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -96,7 +96,9 @@ WarpX::Evolve (int numsteps)
+            if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+                FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
+            }
 #endif
```
Should this call to FillBoundaryM really have to be wrapped in this if statement?  If so that's fine, but in the future if a non-macroscopic option is implemented, one might forget to take this if statement out.

jackieyao0114 replied:
That's true. Maybe separate the cases of "macroscopic" and "magnetic"? So here for FillBoundaryM, instead of using `WarpX::em_solver_medium == MediumForEM::Macroscopic` as the condition, we will use `WarpX::em_solver_medium == MediumForEM::Magnetic`?

ajnonaka replied:
I'm just wondering if there is ever a case where simply always calling FillBoundaryM will hurt anything.
I figured the #ifdef WARPX_MAG_LLG being true was enough to say we cared about M field evolution.

jackieyao0114 replied:
Agreed. Look at the new commit suggestion that I directly  put here using Reva way :-)

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/EvolveM.cpp:
```diff
@@ -38,6 +38,8 @@ void FiniteDifferenceSolver::EvolveM (
     } */
 
+    amrex::Abort("The function EvolveM should not be used for macroscopic materials");
+
```
Shouldn't this say "non-macroscopic materials"?

jackieyao0114 replied:
This is EvolveM.cpp, which is meant to be used for non-macroscopic materials.

ajnonaka commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -252,12 +252,19 @@ WarpX::InitLevelData (int lev, Real /*time*/)
                    ::tolower);
-
+# ifdef WARPX_MAG_LLG
     pp.query("M_ext_grid_init_style", M_ext_grid_s); // user-defined initial M
```
"#ifdef" not "# ifdef"

jackieyao0114 replied:
fixed

jackieyao0114 committed:
fixed the space in '# ifdef' in Source/Initialization/WarpXInitData.cpp

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H:
```diff
@@ -81,6 +81,14 @@ class FiniteDifferenceSolver
                             amrex::Real const dt,
                             std::unique_ptr<MacroscopicProperties> const& macroscopic_properties);
 
```
Can you please add a doxygen style comment here describing very briefly what theMacroscopicEvolveM function does and what each input parameter is? See the doxygen style comment for MacroscopicEvolveE above for reference

jackieyao0114 replied:
Added. Thank you!

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,236 @@
+    else
+    {
+        amrex::Abort("Unknown algorithm");
+    } */
```
```suggestion

```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,236 @@
+        // static constexpr amrex::Real alpha = 1e-4;
+        // static constexpr amrex::Real Ms = 1e4;
+        // Real constexpr cons1 = -mag_gamma_interp; // should be mu0*gamma, mu0 is absorbed by B used in this case
+        // Real constexpr cons2 = -cons1*alpha/Ms; // factor of the second term in scalar LLG
```
```suggestion

```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,236 @@
+          auto& mag_Ms_mf = macroscopic_properties->getmag_Ms_mf();
+          auto& mag_alpha_mf = macroscopic_properties->getmag_alpha_mf();
+          auto& mag_gamma_mf = macroscopic_properties->getmag_gamma_mf();
+          // exctract material properties
```
```suggestion
          // extract material properties
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,236 @@
+        // Real constexpr cons2 = -cons1*alpha/Ms; // factor of the second term in scalar LLG
+
+        for (MFIter mfi(*Mfield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi) /* remember to FIX */
+        {
```
could you please ensure that there are 4 white spaces after the curly braces within the loop

jackieyao0114 replied:
The indention is fixed.

jackieyao0114 committed:
clear up EvolveM, leaving only MacroscopicEvolveM

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,219 @@
+              // Real Hx_eff = Hx_xface + Hx_bias_xface;
+              // Real Hy_eff = Hy_xface + Hy_bias_xface;
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
+
```
Do you want to uncomment these lines to generalize the code now?
You could also save memory by saying:

Hx_xface += Hx_bias_xface;

or even save MORE memory by saying:

Hx_xface += MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Hx_bias);

and then using Hx_xface instead of Hx_eff below.
Same comments for the y and z loops.

jackieyao0114 replied:
Let's add H_Maxwell back in the next PR after we approve EvolveM_HtoE from Jordan. Without the correct way of calculating E using H, adding H_Maxwell to LLG will not give the correct results anyways...
And yes, I like the way you suggested to save the memory :-)

RevathiJambunathan replied:
so what about the comments out lines 102 to 104? if not required, we can delete them. if we are just not using them yet, may be add a comment there to ensure that it is uncommented in the future

jackieyao0114 replied:
I have added a comment before line 102 to remind us to add H_Maxwell back.

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+         i,j,k are the spatial indices, n is the component index
+         iv_in is the input field index type, iv_out is the target index type
+     */
+
```
You should delete large commented out sections of the code unless they are relevant to explaining what is going on.  This looks like old test code?

ajnonaka commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -420,10 +441,111 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+                                                    lev);
+       }
+    }
+    */
```
What's going on here?  Does the parser code actually work if it were uncommented?

jackieyao0114 replied:
Parser will be the next PR. So I've deleted the commented code here.  Thanks for catching this~

ajnonaka commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -96,7 +96,9 @@ WarpX::Evolve (int numsteps)
-            FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
+            if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+                FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
+            }
```
Wondering if you should remove the "if" test in the possible case there is a non-macroscopic version of evolveM someday (and you don't want to have to remember to undo all these "if" tests).  You have aborts elsewhere preventing non-macroscopic evolveM from running anyway, so calling FillBoundaryM no matter what makes the code cleaner without changing the functionality.

ajnonaka reviewed (approved):
I'm not going to check the details of all the spatial discretizations (averaging, curls, etc.) - that is what the convergence and comparison testing is for.  But as a whole, structurally, I think the PR looks good to merge.  I'll still be looking and adding cleanup comments until it's merged.

ajnonaka commented on Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM:
```diff
@@ -0,0 +1,92 @@
+macroscopic.mag_alpha_function(x,y,z) = "5e-01"
+#macroscopic.mag_alpha = 5e-02 # alpha is unitless, typical values range from 1e-3 ~ 1e-5
+macroscopic.mag_gamma_init_style = "parse_mag_gamma_function"
+macroscopic.mag_gamma_function(x,y,z) = "1.759e11"
```
per discussion, gamma should be negative.  alpha is positive.

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+                             + macro_mag_prop(i-iv[0]+1,j-iv[1]+1,k-iv[2]+1)
+         ));
+     };
+     /*  Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
+         Hz (0,0,1) components to Hx face (1,0,0)
+         Real Hy_xtemp = 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
+         Real Hz_xtemp = 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
+
+         Hx (1,0,0) components to Hy face (0,1,0)
+         Hz (0,0,1) components to Hy face (0,1,0)
+         Real Hx_ytemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i,j-1,k)+Hx(i+1,j-1,k));
+         Real Hz_ytemp = 0.25*(Hz(i,j,k)+Hz(i,j,k+1)+Hz(i,j-1,k)+Hz(i,j-1,k+1));
+       
+         Hx (1,0,0) components to Hz face (0,0,1)
+         Hy (0,1,0) components to Hz face (0,0,1)
+
+         Real Hx_ztemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i+1,j,k-1)+Hx(i,j,k-1));
+         Real Hy_ztemp = 0.25*(Hy(i,j,k)+Hy(i,j,k-1)+Hy(i,j+1,k)+Hy(i,j+1,k-1));
+
+         i,j,k are the spatial indices, n is the component index
+         iv_in is the input field index type, iv_out is the target index type
+     */
```
```suggestion
```

ajnonaka replied:
```suggestion
```

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+                             + macro_mag_prop(i-iv[0]+1,j-iv[1]+1,k-iv[2]+1)
+         ));
+     };
+     /*  Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
+         Hz (0,0,1) components to Hx face (1,0,0)
+         Real Hy_xtemp = 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
+         Real Hz_xtemp = 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
+
+         Hx (1,0,0) components to Hy face (0,1,0)
+         Hz (0,0,1) components to Hy face (0,1,0)
+         Real Hx_ytemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i,j-1,k)+Hx(i+1,j-1,k));
+         Real Hz_ytemp = 0.25*(Hz(i,j,k)+Hz(i,j,k+1)+Hz(i,j-1,k)+Hz(i,j-1,k+1));
+       
+         Hx (1,0,0) components to Hz face (0,0,1)
+         Hy (0,1,0) components to Hz face (0,0,1)
+
+         Real Hx_ztemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i+1,j,k-1)+Hx(i,j,k-1));
+         Real Hy_ztemp = 0.25*(Hy(i,j,k)+Hy(i,j,k-1)+Hy(i,j+1,k)+Hy(i,j+1,k-1));
+
+         i,j,k are the spatial indices, n is the component index
+         iv_in is the input field index type, iv_out is the target index type
+     */
```
```suggestion
```

ajnonaka replied:
(delete commented out code)

RevathiJambunathan commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -420,10 +441,111 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+                                                    str_Mx_ext_grid_function);
+       Store_parserString(pp, "My_external_grid_function(x,y,z)",
+                                                    str_My_ext_grid_function);
+       Store_parserString(pp, "Mz_external_grid_function(x,y,z)",
+                                                    str_Mz_ext_grid_function);
+
+       Mxfield_parser.reset(new ParserWrapper<3>(
+                                makeParser(str_Mx_ext_grid_function,{"x","y","z"})));
+       Myfield_parser.reset(new ParserWrapper<3>(
+                                makeParser(str_My_ext_grid_function,{"x","y","z"})));
+       Mzfield_parser.reset(new ParserWrapper<3>(
+                                makeParser(str_Mz_ext_grid_function,{"x","y","z"})));
+
+       // Initialize Mfield_fp with external function
+       InitializeExternalFieldsOnGridUsingParser(Mfield_fp[lev][0].get(),
+                                                 Mfield_fp[lev][1].get(),
+                                                 Mfield_fp[lev][2].get(),
+                                                 Mxfield_parser.get(),
+                                                 Myfield_parser.get(),
+                                                 Mzfield_parser.get(),
+                                                 lev);
+       if (lev > 0) {
+          InitializeExternalFieldsOnGridUsingParser(Mfield_aux[lev][0].get(),
+                                                    Mfield_aux[lev][1].get(),
+                                                    Mfield_aux[lev][2].get(),
+                                                    Mxfield_parser.get(),
+                                                    Myfield_parser.get(),
+                                                    Mzfield_parser.get(),
+                                                    lev);
+
+          InitializeExternalFieldsOnGridUsingParser(Mfield_cp[lev][0].get(),
+                                                    Mfield_cp[lev][1].get(),
+                                                    Mfield_cp[lev][2].get(),
+                                                    Mxfield_parser.get(),
+                                                    Myfield_parser.get(),
+                                                    Mzfield_parser.get(),
+                                                    lev);
+       }
+    }
+    */
```
```suggestion
```

jackieyao0114 replied:
```suggestion
```

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+                             + macro_mag_prop(i-iv[0]+1,j-iv[1]+1,k        )
+                             + macro_mag_prop(i-iv[0]+1,j-iv[1]+1,k-iv[2]+1)
+         ));
+     };
+     /*  Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
+         Hz (0,0,1) components to Hx face (1,0,0)
+         Real Hy_xtemp = 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
+         Real Hz_xtemp = 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
+
+         Hx (1,0,0) components to Hy face (0,1,0)
+         Hz (0,0,1) components to Hy face (0,1,0)
+         Real Hx_ytemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i,j-1,k)+Hx(i+1,j-1,k));
+         Real Hz_ytemp = 0.25*(Hz(i,j,k)+Hz(i,j,k+1)+Hz(i,j-1,k)+Hz(i,j-1,k+1));
+       
+         Hx (1,0,0) components to Hz face (0,0,1)
+         Hy (0,1,0) components to Hz face (0,0,1)
+
+         Real Hx_ztemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i+1,j,k-1)+Hx(i,j,k-1));
+         Real Hy_ztemp = 0.25*(Hy(i,j,k)+Hy(i,j,k-1)+Hy(i,j+1,k)+Hy(i,j+1,k-1));
+
+         i,j,k are the spatial indices, n is the component index
+         iv_in is the input field index type, iv_out is the target index type
+     */
+
+     AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```
```suggestion
     };
     
     /*  Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
         Hz (0,0,1) components to Hx face (1,0,0)
         Real Hy_xtemp = 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
         Real Hz_xtemp = 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));

         Hx (1,0,0) components to Hy face (0,1,0)
         Hz (0,0,1) components to Hy face (0,1,0)
         Real Hx_ytemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i,j-1,k)+Hx(i+1,j-1,k));
         Real Hz_ytemp = 0.25*(Hz(i,j,k)+Hz(i,j,k+1)+Hz(i,j-1,k)+Hz(i,j-1,k+1));
       
         Hx (1,0,0) components to Hz face (0,0,1)
         Hy (0,1,0) components to Hz face (0,0,1)

         Real Hx_ztemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i+1,j,k-1)+Hx(i,j,k-1));
         Real Hy_ztemp = 0.25*(Hy(i,j,k)+Hy(i,j,k-1)+Hy(i,j+1,k)+Hy(i,j+1,k-1));

         i,j,k are the spatial indices, n is the component index
         iv_in is the input field index type, iv_out is the target index type
     */
     AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

jackieyao0114 committed:
fixed signs of mag_gamma and mag_alpha; added Abort to user-input values that violate the rule of signs

jackieyao0114 commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -96,7 +96,9 @@ WarpX::Evolve (int numsteps)
             FillBoundaryB(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
 #ifdef WARPX_MAG_LLG
-            FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
+            if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+                FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
+            }
```
```suggestion
            FillBoundaryM(guard_cells.ng_alloc_EB, guard_cells.ng_Extra);
```

ajnonaka replied:
I agree with this change.

ajnonaka replied:
I agree with this change.  I think Jackie is the one who accepts this?

jackieyao0114 commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -115,7 +117,9 @@ WarpX::Evolve (int numsteps)
             FillBoundaryB(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
 #ifdef WARPX_MAG_LLG
-            FillBoundaryM(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
+            if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+                FillBoundaryM(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
+            }
```
```suggestion
            FillBoundaryM(guard_cells.ng_FieldGather, guard_cells.ng_Extra);
```

ajnonaka replied:
good

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+     amrex::MultiFab& getmag_alpha_mf () {return (*m_mag_alpha_mf);}
+     amrex::MultiFab& getmag_gamma_mf () {return (*m_mag_gamma_mf);}
+
+     // interpolate the magnetic properties to B locations
+     // magnetic properties are cell nodal
+     // B locations are face centered
```
```suggestion
     // interpolate the magnetic properties to B locations
     // magnetic properties are cell nodal
     // B locations are face centered
     // iv is an IntVect with a 1 in the face direction of interest, and 0 in the others
```

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+         ));
+     };
+     
+     /*  Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
+         Hz (0,0,1) components to Hx face (1,0,0)
+         Real Hy_xtemp = 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
+         Real Hz_xtemp = 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
+
+         Hx (1,0,0) components to Hy face (0,1,0)
+         Hz (0,0,1) components to Hy face (0,1,0)
+         Real Hx_ytemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i,j-1,k)+Hx(i+1,j-1,k));
+         Real Hz_ytemp = 0.25*(Hz(i,j,k)+Hz(i,j,k+1)+Hz(i,j-1,k)+Hz(i,j-1,k+1));
+       
+         Hx (1,0,0) components to Hz face (0,0,1)
+         Hy (0,1,0) components to Hz face (0,0,1)
+
+         Real Hx_ztemp = 0.25*(Hx(i,j,k)+Hx(i+1,j,k)+Hx(i+1,j,k-1)+Hx(i,j,k-1));
+         Real Hy_ztemp = 0.25*(Hy(i,j,k)+Hy(i,j,k-1)+Hy(i,j+1,k)+Hy(i,j+1,k-1));
+
+         i,j,k are the spatial indices, n is the component index
+         iv_in is the input field index type, iv_out is the target index type
+     */
```
```suggestion
     /*  
         This routine averages a face-centered field onto a (possibly different nodality) face.
         i,j,k are the spatial indices, n is the component index
         iv_in is the nodality of the input field, iv_out is the nodality of the output.
         Thus, if iv_in is the same as iv_out the value on the face is returned.
         Some other examples when iv_in is different from iv_out:
         
         Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
         return 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));

         Hz (0,0,1) components to Hx face (1,0,0)
         return 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
     */
```

ajnonaka commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,144 @@ public:
+                             )); 
+     };
+
+     /** 
+     calculate local H_maxwell
+     note Bcomp always has n=0 because it is a 3d matrix
+     while Mcomp has n=0,1,2, indicating x,y,z components
+     **/
```
```suggestion
     /*
     calculate local H_maxwell = B / mu - M
     n is the component of the H field we are interested in
     iv_in is the nodality of the input B field, Bcomp
     iv_out is the nodality of the output
     note we wall face_avg_to_face on the B field with n=0 (B has only 1 component)
     while Mcomp has n=0,1,2, indicating x,y,z components
     note Mcomp has the same nodality as the final output H_maxwell
     */
```

ajnonaka reviewed (approved):
pending my comments, I think this is fine to merge.

jackieyao0114 committed:
Update Source/Evolve/WarpXEvolve.cpp

removed the unnecessary condition "WarpX::em_solver_medium == MediumForEM::Macroscopic" and keep only ifdef LLG

jackieyao0114 committed:
Update Source/Evolve/WarpXEvolve.cpp

removed the unnecessary condition "WarpX::em_solver_medium == MediumForEM::Macroscopic" and keep only ifdef LLG

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

added comments on macro_avg_to_face for better understanding

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

added more comments in face_avg_to_face for better understanding

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

added more comments on getH_Maxwell for better understanding

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,219 @@
+// update M field over one timestep
+
+void FiniteDifferenceSolver::MacroscopicEvolveM (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Mfield, // Mfield contains three components MultiFab
```
The comment needs to be slightly re-worded.
The MField here is a vector of three multifabs, with M on each face, and each multifab is a three-component multifab.

jackieyao0114 replied:
updated. Thank you!

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+         ));
+     };
+     
+     /*  
```
```suggestion
     /** \brief
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+     };
+     
+     /*  
+         This routine averages a face-centered field onto a (possibly different nodality) face.
```
```suggestion
         This GPU kernel averages a face-centered field onto a (possibly different nodality) face.
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+                             )); 
+     };
+
+     /*
```
```suggestion
     /** \brief
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+     };
+
+     /*
+     calculate local H_maxwell = B / mu - M
```
```suggestion
     Calculate local H_maxwell = B / mu_o - M
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+     n is the component of the H field we are interested in
+     iv_in is the nodality of the input B field, Bcomp
+     iv_out is the nodality of the output
+     note we wall face_avg_to_face on the B field with n=0 (B has only 1 component)
```
```suggestion
     Note, we call face_avg_to_face on the B field with n=0 (B has only 1 component)
```

RevathiJambunathan replied:
Can you add doxygen style comments here as well along with \param[in]

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+         return 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
+
+         Hz (0,0,1) components to Hx face (1,0,0)
+         return 0.25*(Hz(i,j,k)+Hz(i-1,j,k)+Hz(i,j,k+1)+Hz(i-1,j,k+1));
```
```suggestion
         \param[in] Fieldcomp  floating point data to be interpolated from one staggering location to another.
         \param[in] iv_in           nodality of the input field, Fieldcomp
         \param[in] iv_out         nodality of the output
         \param[in] i                index along x of the Array4 Fieldcomp
         \param[in] j                index along y of the Array4 Fieldcomp
         \param[in] k                index along z of the Array4 Fieldcomp
```

RevathiJambunathan replied:
Inlcuding doc is  /** brief 
                           */
will document the code in doxygen style, which will automatically update the doc webpage when we create one.

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
      /** Multifab for m_mu */
      std::unique_ptr<amrex::MultiFab> m_mu_mf;
+#ifdef WARPX_MAG_LLG
+     /** Multifab for m_mag_Ms */
```
```suggestion
     /** Multifab storing spatially varying saturation magnetization */
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+#ifdef WARPX_MAG_LLG
+     /** Multifab for m_mag_Ms */
+     std::unique_ptr<amrex::MultiFab> m_mag_Ms_mf;
+     /** Multifab for m_mag_alpha */
```
```suggestion
     /** Multifab storing spatially varying Gilbert damping */
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,143 @@ public:
+     std::unique_ptr<amrex::MultiFab> m_mag_Ms_mf;
+     /** Multifab for m_mag_alpha */
+     std::unique_ptr<amrex::MultiFab> m_mag_alpha_mf;
+     /** Multifab for m_mag_gamma */
```
```suggestion
     /** Multifab storing spatially varying gyromagnetic ratio */
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -43,6 +43,37 @@ MacroscopicProperties::ReadParameters ()
     }
 
+#ifdef WARPX_MAG_LLG
+    pp.get("mag_Ms_init_style", m_mag_Ms_s);
```
Can you add documentation for all input parameter that are parsed here in WarpX/docs/running_cpp/parameters.rst.
I am not able to access that particular file from the GUI. 
I can show you how to do this offline.

jackieyao0114 replied:
Sure. Thank you!

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -43,6 +43,37 @@ MacroscopicProperties::ReadParameters ()
+
+    pp.get("mag_alpha_init_style", m_mag_alpha_s);
+    if (m_mag_alpha_s == "constant") pp.get("mag_alpha", m_mag_alpha); 
+    // _mag_ such that it's clear the Ms variable is only meaningful for magnetic materials
```
```suggestion
    // _mag_ such that it's clear the alpha variable is only meaningful for magnetic materials
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -43,6 +43,37 @@ MacroscopicProperties::ReadParameters ()
+
+    pp.get("mag_gamma_init_style", m_mag_gamma_s);
+    if (m_mag_gamma_s == "constant") pp.get("mag_gamma", m_mag_gamma); 
+    // _mag_ such that it's clear the Ms variable is only meaningful for magnetic materials
+    //initialization with parser
```
```suggestion
    // _mag_ such that it's clear the gamma variable parsed here is only meaningful for magnetic materials
    //initialization with parser
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -61,6 +92,12 @@ MacroscopicProperties::InitData ()
     m_eps_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
     m_mu_mf = std::make_unique<MultiFab>(amrex::convert(ba,amrex::IntVect::TheUnitVector()), dmap, 1, ng);
 
+#ifdef WARPX_MAG_LLG
+    m_mag_Ms_mf = std::make_unique<MultiFab>(amrex::convert(ba,amrex::IntVect::TheUnitVector()), dmap, 1, ng);
```
```suggestion
#ifdef WARPX_MAG_LLG
    // all magnetic macroparameters are stored on cell nodes
    m_mag_Ms_mf = std::make_unique<MultiFab>(amrex::convert(ba,amrex::IntVect::TheUnitVector()), dmap, 1, ng);
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,219 @@
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
```
```suggestion
              // keep the interpolation. The IntVect is (1,0,0) to interpolate values to the x-face.
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -90,6 +127,38 @@ MacroscopicProperties::InitData ()
+        InitializeMacroMultiFabUsingParser(m_mag_alpha_mf.get(), m_mag_alpha_parser.get(), lev);
+    }
+    if (m_mag_alpha_mf->min(0,m_mag_alpha_mf->nGrow()) < 0) {
+        Abort("alpha should be positive, but the user input has negative values");
```
```suggestion
        amrex::Abort("alpha should be positive, but the user input has negative values");
```

RevathiJambunathan commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:
```diff
@@ -90,6 +127,38 @@ MacroscopicProperties::InitData ()
+        InitializeMacroMultiFabUsingParser(m_mag_gamma_mf.get(), m_mag_gamma_parser.get(), lev);
+    }
+    if (m_mag_gamma_mf->max(0,m_mag_gamma_mf->nGrow()) > 0) {
+        Abort("gamma should be negative, but the user input has positive values");
```
```suggestion
        amrex::Abort("gamma should be negative, but the user input has positive values");
```

RevathiJambunathan commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -420,10 +441,111 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+    // if the input string for the Hbias-field is "parse_h_bias_ext_grid_function",
+    // then the analytical expression or function must be
+    // provided in the input file.
+    if (H_bias_ext_grid_s == "parse_h_bias_ext_grid_function") {
```
add this input option in the docs as well. WarpX/docs/running_cpp/parameters.rst

jackieyao0114 replied:
Sure. Let's talk offline

RevathiJambunathan commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -420,10 +441,111 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+#ifdef WARPX_DIM_RZ
+       amrex::Abort("H bias parser for external fields does not work with RZ -- TO DO");
+#endif
+       Store_parserString(pp, "Hx_bias_external_grid_function(x,y,z)",
```
these input options too

RevathiJambunathan commented on Source/WarpX.H:
```diff
@@ -103,13 +103,21 @@ public:
+
+#ifdef WARPX_MAG_LLG
     static amrex::Vector<amrex::Real> M_external_grid;
+    static amrex::Vector<amrex::Real> H_bias_external_grid;
```
can you delete the commented our M_external_grid definition below, on line 110

jackieyao0114 replied:
deleted.

RevathiJambunathan commented on Source/WarpX.H:
```diff
@@ -666,7 +708,7 @@ private:
 
     void ExchangeWithPmlB (int lev);
-    void ExchangeWithPmlE (int lev); // does M need this??
+    void ExchangeWithPmlE (int lev); // do M & H need this???
```
Do we know the answer to this comment?

jackieyao0114 replied:
I think M and H_Maxwell do need this function. Correct me if I'm wrong...

RevathiJambunathan commented on Source/WarpX.H:
```diff
@@ -322,18 +362,21 @@ public:
+    void MacroscopicEvolveM (int lev, amrex::Real dt);
+    void MacroscopicEvolveM (int lev, PatchType patch_type, amrex::Real dt);
+#endif
+    void FieldGather ();
```
This FieldGather () function is not there in Source/WarpX.H and not part of the PR. this can be deleted. probably a left over from some merge?

jackieyao0114 replied:
I've deleted it. Thank you for catching this.

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -992,6 +1051,10 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
         Mfield_aux[lev][0].reset( new MultiFab(amrex::convert(ba,Mx_nodal_flag),dm,3     ,ngE));
         Mfield_aux[lev][1].reset( new MultiFab(amrex::convert(ba,My_nodal_flag),dm,3     ,ngE));
         Mfield_aux[lev][2].reset( new MultiFab(amrex::convert(ba,Mz_nodal_flag),dm,3     ,ngE));
+
+        H_biasfield_aux[lev][0].reset( new MultiFab(amrex::convert(ba,Hx_bias_nodal_flag),dm,ncomps,ngE));
+        H_biasfield_aux[lev][1].reset( new MultiFab(amrex::convert(ba,Hy_bias_nodal_flag),dm,ncomps,ngE));
+        H_biasfield_aux[lev][2].reset( new MultiFab(amrex::convert(ba,Hz_bias_nodal_flag),dm,ncomps,ngE));
```
all this is in ifdef PSATD. do we need to define these multifabs in the PSATD scope? we are not using PSATD and dont support evolveM for the method. may be removing it from here would be best

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -967,15 +1015,26 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
         Efield_aux[lev][0].reset( new MultiFab(nba,dm,ncomps,ngE));
         Efield_aux[lev][1].reset( new MultiFab(nba,dm,ncomps,ngE));
         Efield_aux[lev][2].reset( new MultiFab(nba,dm,ncomps,ngE));
+#ifdef WARPX_MAG_LLG
+        Mfield_aux[lev][0].reset( new MultiFab(nba,dm,3     ,ngE));
+        Mfield_aux[lev][1].reset( new MultiFab(nba,dm,3     ,ngE));
+        Mfield_aux[lev][2].reset( new MultiFab(nba,dm,3     ,ngE));
+
+        H_biasfield_aux[lev][0].reset( new MultiFab(nba,dm,ncomps,ngE));
+        H_biasfield_aux[lev][1].reset( new MultiFab(nba,dm,ncomps,ngE));
+        H_biasfield_aux[lev][2].reset( new MultiFab(nba,dm,ncomps,ngE));
+#endif
```
all this is in ifdef PSATD. May not be required?

jackieyao0114 replied:
hmmm. I have hard time recognizing the end of #ifdef WARPX_USE_PSATD in line 968. Maybe you can help me with it?

jackieyao0114 replied:
PSATD does not cover this part per our discussion. I will close this conversation.

RevathiJambunathan reviewed (changes requested):
Thank you for this PR, @jackieyao0114 !!!
It looks great! 

Please see my comments below. 
One of my main comments is to include input paramters, i.e., user-defined input options in the docs. 
This can be added in WarpX/docs/running_cpp/parameters.rst.

Please let me know if you have any questions.

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

deleted unnecessary comments in MacroscopicEvolveM.cpp

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H:
```diff
@@ -81,6 +74,14 @@ class FiniteDifferenceSolver
+                       amrex::Real const dt,
+                       std::unique_ptr<MacroscopicProperties> const& macroscopic_properties);
+#endif
+
```
```suggestion
#ifdef WARPX_MAG_LLG
        /**
          * \brief Macroscopic M-update for non-vacuum medium using finite-difference algorithm 
          * solving Landau-Lifshitz-Gilbert (LLG) equation,
          * only Yee's algorithm is applicable for M calculation
          * This function has first-order accuracy with forward-Euler method 
          * \param[out] Mfield  vector of magnetization MultiFabs updated at a given level; each MultiFab locates
          * on the face centers of the spatial cell; and each MultiFab contains three four-dimensional FabArrays
          * indicating the x, y, z locations and the field component
          * \param[in] H_biasfield   vector of user-defined DC magnetic bias field MultiFabs at a given level
          * \param[in] Bfield   vector of magnetic field MultiFabs at a given level
          * \param[in] dt       timestep of the simulation
          * \param[in] macroscopic_properties contains user-defined properties of the medium.
          */
          
        void MacroscopicEvolveM ( std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Mfield,
                       std::array< std::unique_ptr<amrex::MultiFab>, 3 >& H_biasfield,
                       std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
                       amrex::Real const dt,
                       std::unique_ptr<MacroscopicProperties> const& macroscopic_properties);
#endif

```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

fix typo

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+// update M field over one timestep
+
+void FiniteDifferenceSolver::MacroscopicEvolveM (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Mfield, // Mfield contains three components MultiFab
```
```suggestion
    // The MField here is a vector of three multifabs, with M on each face, and each multifab is a three-component multifab. 
    // Each M-multifab has three components, one for each component in x, y, z. (All multifabs are four dimensional, (i,j,k,n)), where, n=1 for E, B, but, n=3 for M_xface, M_yface, M_zface
    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Mfield, 
```

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+          Array4<Real> const& mag_Ms_arr = mag_Ms_mf.array(mfi);
+          Array4<Real> const& mag_alpha_arr = mag_alpha_mf.array(mfi);
+          Array4<Real> const& mag_gamma_arr = mag_gamma_mf.array(mfi);
+
```
```suggestion
        for (MFIter mfi(*Mfield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi) /* remember to FIX */
        {
            auto& mag_Ms_mf = macroscopic_properties->getmag_Ms_mf();
            auto& mag_alpha_mf = macroscopic_properties->getmag_alpha_mf();
            auto& mag_gamma_mf = macroscopic_properties->getmag_gamma_mf();
            // extract material properties
            Array4<Real> const& mag_Ms_arr = mag_Ms_mf.array(mfi);
            Array4<Real> const& mag_alpha_arr = mag_alpha_mf.array(mfi);
            Array4<Real> const& mag_gamma_arr = mag_gamma_mf.array(mfi);
            
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

completed comments of interpolating for better understanding

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+              Real Hy_eff = Hy_bias_xface;
+              Real Hz_eff = Hz_bias_xface;
+
+              // Real Hx_eff = Hx_xface + Hx_bias_xface;
+              // Real Hy_eff = Hy_xface + Hy_bias_xface;
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
```
```suggestion
              // keep in mind to ADD H_Maxwell back to H_eff as shown below
              // Real Hx_eff = Hx_xface + Hx_bias_xface;
              // Real Hy_eff = Hy_xface + Hy_bias_xface;
              // Real Hz_eff = Hz_xface + Hz_bias_xface;
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

added doxygen style comment \brief to face_ave_to_face function

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+              Real Hy_bias_yface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,1,0), amrex::IntVect(0,1,0), Hy_bias);
+              Real Hz_bias_yface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,0,1), amrex::IntVect(0,1,0), Hz_bias);
+              // H_eff = H_maxwell + H_bias + H_exchange + H_anisotropy ... (only the first two terms are considered here)
+              // Real Hx_eff = Hx_yface + Hx_bias_yface;
+              // Real Hy_eff = Hy_yface + Hy_bias_yface;
+              // Real Hz_eff = Hz_yface + Hz_bias_yface;
```
```suggestion
              // keep in mind to ADD H_Maxwell back to H_eff as shown below
              // Real Hx_eff = Hx_yface + Hx_bias_yface;
              // Real Hy_eff = Hy_yface + Hy_bias_yface;
              // Real Hz_eff = Hz_yface + Hz_bias_yface;
```

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+              Real Hz_eff = Hz_bias_yface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
```
```suggestion
              // keep the interpolation. The IntVect is (0,1,0) to interpolate values to the y-face.
```

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+              Real Hy_bias_zface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,1,0), amrex::IntVect(0,0,1), Hy_bias);
+              Real Hz_bias_zface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,0,1), amrex::IntVect(0,0,1), Hz_bias);
+              // H_eff = H_maxwell + H_bias + H_exchange + H_anisotropy ... (only the first two terms are considered here)
+              // Real Hx_eff = Hx_zface + Hx_bias_zface;
+              // Real Hy_eff = Hy_zface + Hy_bias_zface;
+              // Real Hz_eff = Hz_zface + Hz_bias_zface;
```
```suggestion
              // keep in mind to ADD H_Maxwell back to H_eff as shown below
              // Real Hx_eff = Hx_zface + Hx_bias_zface;
              // Real Hy_eff = Hy_zface + Hy_bias_zface;
              // Real Hz_eff = Hz_zface + Hz_bias_zface;
```

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp:
```diff
@@ -0,0 +1,215 @@
+              Real Hz_eff = Hz_bias_zface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
```
```suggestion
              // keep the interpolation. The IntVect is (0,0,1) to interpolate values to the z-face.
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

change the wording of comment to be more accurate. "routine" --> "GPU kernel"

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

added doxygen comment to face_ave_to_face

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,148 @@ public:
+         \param[in] i                index along x of the Array4 Fieldcomp
+         \param[in] j                index along y of the Array4 Fieldcomp
+         \param[in] k                index along z of the Array4 Fieldcomp
+     */
```
```suggestion
     /** \brief
         * This GPU kernel averages a face-centered field onto a (possibly different nodality) face.
         * i,j,k are the spatial indices, n is the component index
         * iv_in is the nodality of the input field, iv_out is the nodality of the output.
         * Thus, if iv_in is the same as iv_out the value on the face is returned.
         * Some other examples when iv_in is different from iv_out:
         
         * Hy (0,1,0) components (in) to Hx face (1,0,0) (out)
         * return 0.25*(Hy(i,j,k)+Hy(i,j+1,k)+Hy(i-1,j,k)+Hy(i-1,j+1,k));
         * 
         * \param[in] Fieldcomp  floating point data to be interpolated from one staggering location to another.
         * \param[in] iv_in           nodality of the input field, Fieldcomp
         * \param[in] iv_out         nodality of the output
         * \param[in] i                index along x of the Array4 Fieldcomp
         * \param[in] j                index along y of the Array4 Fieldcomp
         * \param[in] k                index along z of the Array4 Fieldcomp
     */
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

fixed constitutive relation H_Maxwell = B/mu0 -M

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

fixed typos in comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

fixed comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

fixed comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H

fixed comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,148 @@ public:
+     Note, we call face_avg_to_face on the B field with n=0 (B has only 1 component)
+     while Mcomp has n=0,1,2, indicating x,y,z components
+     note Mcomp has the same nodality as the final output H_maxwell
+     */
```
```suggestion
     /** \brief
     * Calculate local H_maxwell = B / mu_o - M
     * \param[out] H_Maxwell local H_Maxwell field calculated using the magnetic constitutive relation
     * H_Maxwell is a scalar
     * \param[in] i x-index of the location under consideration
     * \param[in] j y-index of the location under consideration
     * \param[in] k z-index of the location under consideration
     * \param[in] n x, y, and z component of the H field we are interested in
     * \param[in] iv_in nodality of the input B field, Bcomp
     * \param[in] iv_out nodality of the output H_Maxwell
     * Note, we call face_avg_to_face on the B field with n=0 (B has only 1 component)
     * while Mcomp has n=0,1,2, indicating x,y,z components
     * note Mcomp has the same nodality as the final output H_maxwell
     */
```

jackieyao0114 commented on Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.H:
```diff
@@ -41,32 +41,148 @@ public:
+     amrex::Real m_mag_Ms;
+     /** Gilbert damping, only applies for magnetic materials */
+     amrex::Real m_mag_alpha;
+     /** gyromagnetic ratio, only applies for magnetic materials */
```
```suggestion
     /** gyromagnetic ratio, should be a negative value, only applies for magnetic materials */
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

fixed comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

fixed comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

added comments

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

fixed amrex::Abort

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp

fixed amrex::Abort

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 commented on Source/WarpX.H:
```diff
@@ -103,13 +103,21 @@ public:
+#ifdef WARPX_MAG_LLG
     static amrex::Vector<amrex::Real> M_external_grid;
+    static amrex::Vector<amrex::Real> H_bias_external_grid;
     //static std::array<amrex::Vector<amrex::Real>, 3 > M_external_grid; // define M initial value
```
```suggestion
```

jackieyao0114 commented on Source/WarpX.H:
```diff
@@ -322,18 +362,21 @@ public:
+    void MacroscopicEvolveM (int lev, amrex::Real dt);
+    void MacroscopicEvolveM (int lev, PatchType patch_type, amrex::Real dt);
+#endif
+    void FieldGather ();
```
```suggestion
```

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H

added doxygen comments

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

updated comments

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

fixed indention

jackieyao0114 committed:
Update Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

added notes to H_eff calculation

jackieyao0114 committed:
Apply suggestions from code review

removed redundancy; added comments.

Co-authored-by: Revathi  Jambunathan <41089244+RevathiJambunathan@users.noreply.github.com>

jackieyao0114 committed:
Update Source/WarpX.H

removed redundancy in comments

jackieyao0114 committed:
added parameters related to evolveM to Docs/source/running_cpp/parameters.rst

jackieyao0114 committed:
updated Docs/source/running_cpp/parameters.rst

RevathiJambunathan commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,101 @@ This is essentially the python slicing syntax except that the stop is inclusive
+    requires additional parameters in the input file, namely,
+    ``macroscopic.mag_gamma_function(x,y,z)`` to initialize the gyromagnetic ratio.
+    If ``algo.em_solver_medium`` is set to macroscopic, and ``USE_LLG = TRUE``, 
+    then this input property must be provided.
```
<3 Thanks for such a detailed documentation!

RevathiJambunathan commented on Examples/Tests/Macroscopic_Maxwell/inputs_3d.evolveM:
```diff
@@ -0,0 +1,92 @@
+#################################
+warpx.verbose = 0
+warpx.use_filter = 0
+warpx.cfl = 4000
```
just out of curiosity - this cfl > 1 because of the choice of alpha that you mentioned? is the choice of cfl and alpha related.?

jackieyao0114 replied:
nope. CFL has nothing to do with alpha... I figured shortly after our meeting that with only LLG, stability is not an issue, so we can choose as large as dt we want (to satisfy Nyquist rule of course)

RevathiJambunathan commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,101 @@ This is essentially the python slicing syntax except that the stop is inclusive
+
+* ``macroscopic.mag_gamma_init_style`` (string) optional (default is "default")
+    This parameter determines the type of initialization for the gyromagnetic ratio
+    of the material. The "default" style initializes the gyromagnetic ratio mag_gamma to 0.0.
```
```suggestion
    of the material. The "default" style initializes the gyromagnetic ratio mag_gamma to 0.0.
    Note that only negative (or zero) values for gyromagnetic ratio are allowed in the simulations and positive values for ``mag_alpha``
```

jackieyao0114 replied:
There is some slight modification to this statement. So i made a separate commit that's similar to your suggestion.

RevathiJambunathan commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,101 @@ This is essentially the python slicing syntax except that the stop is inclusive
+
+* ``macroscopic.mag_alpha_init_style`` (string) optional (default is "default")
+    This parameter determines the type of initialization for the Gilbert damping factor
+    of the material. The "default" style initializes the Gilbert damping factor mag_alpha to 0.0.
```
Note that only positive (or zero) values for Gilbert damping factor are allowed in the simulation.

jackieyao0114 replied:
```suggestion
    of the material. The "default" style initializes the Gilbert damping factor ``mag_alpha`` to 0.0.
    Note that only positive values for Gilbert damping factor ``mag_alpha`` are allowed in the simulation.
```

RevathiJambunathan reviewed (approved):
Thanks @jackieyao0114 for the PR! Looks great! only two minor comments below. will merge after they are included.

RevathiJambunathan commented:
@jackieyao0114 Also, in source/style check, it shows that TABS are included in some documents. CAn you fix them with the commans that are provided in the error message. Thank you :)

jackieyao0114 commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,101 @@ This is essentially the python slicing syntax except that the stop is inclusive
+
+* ``macroscopic.mag_gamma_init_style`` (string) optional (default is "default")
+    This parameter determines the type of initialization for the gyromagnetic ratio
+    of the material. The "default" style initializes the gyromagnetic ratio mag_gamma to 0.0.
```
```suggestion
    of the material. The "default" style initializes the gyromagnetic ratio mag_gamma to 0.0.
    Note that only negative values for gyromagnetic ratio ``mag_gamma`` are allowed in the simulations and positive values for ``mag_alpha`` are allowed. 
    The value of ``mag_gamma`` depends on the type of spins under consideration. For electrons, ``mag_gamma`` is a constant value -1.759e11 Coulomb/kilogram.
```

jackieyao0114 committed:
added mag_alpha and gamma sign constraints in Docs/source/running_cpp/parameters.rst

jackieyao0114 committed:
added mag_gamma sign constraints in Update Docs/source/running_cpp/parameters.rst

ajnonaka commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,104 @@ This is essentially the python slicing syntax except that the stop is inclusive
+    For example, if ``warpx.Mx_external_grid_function(x,y,z)=Mo*x + delta*(y + z)``
+    then the constants `Mo` and `delta` required in the above equation
+    can be set using ``my_constants.Mo=`` and ``my_constants.delta=`` in the
+    input file. For a two-dimensional simulation, it is assumed that the first dimension     is `x` and the second dimension in `z`, and the value of `y` is set to zero.
```
```suggestion
    input file. For a two-dimensional simulation, it is assumed that the first dimension
    is `x` and the second dimension in `z`, and the value of `y` is set to zero.
```

ajnonaka commented on Docs/source/running_cpp/parameters.rst:
```diff
@@ -1741,3 +1741,104 @@ This is essentially the python slicing syntax except that the stop is inclusive
+    For example, if ``warpx.Hx_bias_external_grid_function(x,y,z)=Ho_bias*x + delta*(y + z)``
+    then the constants `Ho_bias` and `delta` required in the above equation
+    can be set using ``my_constants.Ho_bias=`` and ``my_constants.delta=`` in the
+    input file. For a two-dimensional simulation, it is assumed that the first dimension     is `x` and the second dimension in `z`, and the value of `y` is set to zero.
```
```suggestion
    input file. For a two-dimensional simulation, it is assumed that the first dimension
    is `x` and the second dimension in `z`, and the value of `y` is set to zero.
```

jackieyao0114 committed:
fixed tabs in Update Docs/source/running_cpp/parameters.rst

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

jackieyao0114 committed:
fixed tabs in Update Docs/source/running_cpp/parameters.rst

Co-authored-by: Andy Nonaka <AJNonaka@lbl.gov>

jackieyao0114 committed:
fixed tabs

jackieyao0114 committed:
Merge branch 'debug_evolveM_withHeff' of https://github.com/jackieyao0114/WarpX into debug_evolveM_withHeff

jackieyao0114 committed:
fix white space at line end
